File: src/broker-service/BrokerService/src/Entities/Instruments/Instrument.cs
namespace EasyTrade.BrokerService.Entities.Instruments;

public class Instrument(string code, string name, string description)
{
    public int Id { get; set; }
    public int ProductId { get; set; }
    public string Code { get; set; } = code;
    public string Name { get; set; } = name;
    public string Description { get; set; } = description;

    public Instrument(int id, int productId, string code, string name, string description)
        : this(code, name, description)
    {
        Id = id;
        ProductId = productId;
    }
}

File: src/db-adapter/internal/models/instrument.go
package models

import (
	"context"
	"time"
)

type Instrument struct {
	ID          string
	ProductID   string
	Code        string
	Name        string
	Description string
}

type OwnedInstrument struct {
	ID                   string
	AccountID            string
	InstrumentID         string
	Quantity             float64
	LastModificationDate time.Time
}

type InstrumentRepository interface {
	GetByID(ctx context.Context, id string) (*Instrument, error)
	GetAll(ctx context.Context) ([]*Instrument, error)
	GetOwned(ctx context.Context, accountID, instrumentID string) (*OwnedInstrument, error)
	GetAllOwned(ctx context.Context, accountID string) ([]*OwnedInstrument, error)
	AddOwned(ctx context.Context, owned *OwnedInstrument) (*OwnedInstrument, error)
	UpdateOwned(ctx context.Context, owned *OwnedInstrument) (*OwnedInstrument, error)
}
